File: related_scripts/verification/step_F/verify_F_sem.py
"""Step F verification: SEM fitting — compare Python vs MATLAB outputs."""
import sys, struct
import numpy as np
from pathlib import Path
from scipy.spatial.distance import pdist, squareform

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / 'src'))

from caz.sem import _gaussian_weight_matrix, _profile_neg_loglik, fit_sem

# ---------- replicate MATLAB's data loading ----------
ml_dir_D = Path(r'D:\Dataset\DPastCliM-NA\verification\step_D\matlab')
py_dir_A = Path(r'D:\Dataset\DPastCliM-NA\verification\step_A\python')
ml_out   = Path(r'D:\Dataset\DPastCliM-NA\verification\step_F\matlab')
py_out   = Path(r'D:\Dataset\DPastCliM-NA\verification\step_F\python')
py_out.mkdir(parents=True, exist_ok=True)

# yhat from Step D MATLAB
with open(ml_dir_D / 'yhat_full.bin', 'rb') as f:
    T_ml, n_st = struct.unpack('ii', f.read(8))
    yhat = np.frombuffer(f.read(), dtype=np.float32).reshape(T_ml, n_st)

with open(ml_dir_D / 'station_ids.txt') as f:
    station_ids = [l.strip() for l in f if l.strip()]

# obs
import pandas as pd
obs = pd.read_parquet(py_dir_A / 'ghcn_tas_obs.parquet')
obs_jan = obs[(obs['month'] == 1) & (obs['year'] >= 1875) & (obs['year'] <= 1999)]
unique_ids = sorted(obs_jan['ID'].unique())
unique_years = sorted(obs_jan['year'].unique())
id2col = {sid: i for i, sid in enumerate(unique_ids)}
yr2row = {yr: i for i, yr in enumerate(unique_years)}
Y_mat = np.full((len(unique_years), len(unique_ids)), np.nan, dtype=np.float32)
for _, row in obs_jan.iterrows():
    Y_mat[yr2row[row['year']], id2col[row['ID']]] = row['value']

# ESM time alignment
import xarray as xr
ds = xr.open_dataset(r'D:\Dataset\DPastCliM-NA\TraCE21k\TraCE-21K-II.monthly.TREFHT.nc')
time_ka = ds['time'].values.astype(float)
cal_year_frac = 1950.0 + time_ka * 1000.0
months_since_0 = np.round(cal_year_frac * 12).astype(int)
year_arr = months_since_0 // 12
month_arr = months_since_0 % 12 + 1
cal_mask = (year_arr >= 1875) & (year_arr <= 1999) & (month_arr == 1)
years_jan_cal = year_arr[cal_mask]
ds.close()

common_years = sorted(set(unique_years) & set(years_jan_cal))
ia = [unique_years.index(y) for y in common_years]
Y_aligned = Y_mat[ia, :]
valid_count = np.sum(~np.isnan(Y_aligned), axis=0)
keep = valid_count >= 30
keep_ids = [unique_ids[i] for i in range(len(unique_ids)) if keep[i]]

# Match MATLAB intersect behavior: sorted intersection based on station_ids order
sid_set = set(station_ids) & set(keep_ids)
# MATLAB [~,ia2,ib2] = intersect(string(station_ids), unique_ids(keep))
# intersect returns sorted common elements; ia2 indexes into first, ib2 into second
common_sorted = sorted(sid_set)
ia2 = [station_ids.index(s) for s in common_sorted]
ib2 = [unique_ids.index(s) for s in common_sorted]

Y_obs = Y_aligned[:, ib2].T   # (S, T)
Yhat  = yhat[:, ia2].T         # (S, T)
n_stations = Y_obs.shape[0]
T = Y_obs.shape[1]
print(f"Full dataset: {n_stations} stations, {T} years")

residuals = Y_obs - Yhat

# station metadata + Albers projection
meta = pd.read_parquet(py_dir_A / 'ghcn_tas_meta.parquet')
matched_ids = [station_ids[i] for i in ia2]
meta_sub = meta[meta['ID'].isin(matched_ids)].set_index('ID')
st_lat = np.array([meta_sub.loc[sid, 'lat'] for sid in matched_ids])
st_lon = np.array([meta_sub.loc[sid, 'lon'] for sid in matched_ids])

def albers_forward(lon_deg, lat_deg):
    a = 6378137.0; f = 1/298.257222101; e2 = 2*f - f**2; e = np.sqrt(e2)
    phi1, phi2 = np.radians(29.5), np.radians(45.5)
    phi0, lam0 = np.radians(23.0), np.radians(-96.0)
    m1 = np.cos(phi1)/np.sqrt(1-e2*np.sin(phi1)**2)
    m2 = np.cos(phi2)/np.sqrt(1-e2*np.sin(phi2)**2)
    def qfun(p): return (1-e2)*(np.sin(p)/(1-e2*np.sin(p)**2) - np.log((1-e*np.sin(p))/(1+e*np.sin(p)))/(2*e))
    q0, q1, q2 = qfun(phi0), qfun(phi1), qfun(phi2)
    n = (m1**2-m2**2)/(q2-q1); C = m1**2+n*q1; rho0 = a*np.sqrt(C-n*q0)/n
    phi = np.radians(np.asarray(lat_deg, dtype=float))
    lam = np.radians(np.asarray(lon_deg, dtype=float))
    q = qfun(phi); rho = a*np.sqrt(C-n*q)/n; theta = n*(lam-lam0)
    return rho*np.sin(theta), rho0-rho*np.cos(theta)

x_proj, y_proj = albers_forward(st_lon, st_lat)

# Load MATLAB's sub_idx to use the exact same stations
sub_idx_file = ml_out / 'sub_idx.bin'
if sub_idx_file.exists() and n_stations > 500:
    sub_idx = np.fromfile(sub_idx_file, dtype=np.int32) - 1  # MATLAB 1-based
    residuals = residuals[sub_idx, :]
    x_proj = x_proj[sub_idx]
    y_proj = y_proj[sub_idx]
    n_stations = len(sub_idx)
    print(f"Subsampled to {n_stations} stations (using MATLAB indices)")

coords = np.column_stack([x_proj, y_proj])
valid_mask = np.sum(~np.isnan(residuals), axis=1) >= 30

# Compare residuals with MATLAB (no longer overriding — Python runs independently)
ml_res_file = ml_out / 'residuals_sub.bin'
if ml_res_file.exists():
    with open(ml_res_file, 'rb') as f:
        hdr = struct.unpack('ii', f.read(8))
        ml_res = np.frombuffer(f.read(), dtype=np.float32).reshape(hdr[0], hdr[1])
    both_valid = ~np.isnan(residuals) & ~np.isnan(ml_res)
    if both_valid.any():
        res_diff = np.abs(residuals[both_valid] - ml_res[both_valid])
        print(f"Residuals vs MATLAB: max|diff|={res_diff.max():.4e}, mean={res_diff.mean():.4e}, N={both_valid.sum()}")
    print(f"Python NaN frac: {np.isnan(residuals).mean():.4f}, MATLAB: {np.isnan(ml_res).mean():.4f}")

# Recompute valid_mask from actual residuals (may be MATLAB's)
valid_mask = np.sum(~np.isnan(residuals), axis=1) >= 30
print(f"valid_mask sum (final): {valid_mask.sum()}")

# ---------- SEM fitting (same logic as MATLAB) ----------
thresholds = np.linspace(25_000, 100_000, 4)
penalty_weight = 0.1
best_nLL = np.inf
lambda_hat = 0.5
W_best = np.eye(n_stations)
threshold_best = 50_000.0

from scipy.optimize import minimize_scalar

for th in thresholds:
    W = _gaussian_weight_matrix(coords, th)
    result = minimize_scalar(
        lambda lam: _profile_neg_loglik(lam, residuals, W, valid_mask, penalty_weight),
        bounds=(0.0, 0.999), method='bounded',
        options={'xatol': 1e-6, 'maxiter': 200}
    )
    print(f"  threshold={th:.0f}, lambda={result.x:.4f}, nLL={result.fun:.2f}")
    if result.fun < best_nLL:
        best_nLL = result.fun
        lambda_hat = result.x
        W_best = W
        threshold_best = th

print(f"Best: threshold={threshold_best:.0f}, lambda={lambda_hat:.6f}")

# compute eps_mat
A_best = np.eye(n_stations) - lambda_hat * W_best
eps_mat = np.full_like(residuals, np.nan)
for t in range(T):
    res_t = residuals[:, t]
    valid = ~np.isnan(res_t)
    if valid.sum() < 2:
        continue
    eps_mat[valid, t] = A_best[np.ix_(valid, valid)] @ res_t[valid]

sigma2_hat = np.nanmean(eps_mat ** 2, axis=1)

# save Python outputs
with open(py_out / 'sem_params.bin', 'wb') as f:
    f.write(struct.pack('d', lambda_hat))
    f.write(struct.pack('d', threshold_best))
    f.write(struct.pack('ii', n_stations, T))
    f.write(sigma2_hat.astype(np.float64).tobytes())

with open(py_out / 'eps_mat.bin', 'wb') as f:
    f.write(struct.pack('ii', n_stations, T))
    eps_mat.astype(np.float32).tofile(f)

print(f"\nPython outputs saved to: {py_out}")

# ---------- compare with MATLAB ----------
print("\n=== Comparison with MATLAB ===")
with open(ml_out / 'sem_params.bin', 'rb') as f:
    ml_lambda = struct.unpack('d', f.read(8))[0]
    ml_threshold = struct.unpack('d', f.read(8))[0]
    ml_nst, ml_T = struct.unpack('ii', f.read(8))
    ml_sigma2 = np.frombuffer(f.read(), dtype=np.float64)

print(f"MATLAB: lambda={ml_lambda:.6f}, threshold={ml_threshold:.0f}")
print(f"Python: lambda={lambda_hat:.6f}, threshold={threshold_best:.0f}")
print(f"lambda diff: {abs(ml_lambda - lambda_hat):.2e}")

with open(ml_out / 'eps_mat.bin', 'rb') as f:
    hdr = struct.unpack('ii', f.read(8))
    ml_eps = np.frombuffer(f.read(), dtype=np.float32).reshape(hdr[0], hdr[1])

# only compare if same threshold was chosen
if ml_threshold == threshold_best:
    both_valid = ~np.isnan(eps_mat) & ~np.isnan(ml_eps)
    diff = np.abs(eps_mat[both_valid] - ml_eps[both_valid])
    print(f"eps_mat: max|diff|={diff.max():.4e}, mean|diff|={diff.mean():.4e}")
    both_s2_valid = ~np.isnan(sigma2_hat[:ml_sigma2.shape[0]]) & ~np.isnan(ml_sigma2)
    s2_diff = np.abs(sigma2_hat[:ml_sigma2.shape[0]][both_s2_valid] - ml_sigma2[both_s2_valid])
    print(f"sigma2: max|diff|={s2_diff.max():.4e}, mean|diff|={s2_diff.mean():.4e} ({both_s2_valid.sum()} valid stations)")
else:
    print("Different thresholds chosen — comparing raw eps_mat not meaningful")
    print("Check if subsample ordering differs between MATLAB and Python")

File: related_scripts/verification/step_F/verify_F_sem.m
%% Step F verification: SEM fitting on PCR residuals (January, tas)
%  Uses Step C/D calibration residuals to fit SEM, then compares with Python.
%  Outputs: lambda_hat, sigma2_hat, eps_mat, W_best

clear; close all;

%% load Step C/D data and compute residuals
out_dir = 'D:\Dataset\DPastCliM-NA\verification\step_F\matlab';
if ~exist(out_dir,'dir'), mkdir(out_dir); end

% load station predictions and observations from Step C/D
ml_dir_D = 'D:\Dataset\DPastCliM-NA\verification\step_D\matlab';
fid = fopen(fullfile(ml_dir_D, 'yhat_full.bin'), 'r');
hdr = fread(fid, 2, 'int32');
T = hdr(1); n_st = hdr(2);
yhat = fread(fid, [n_st, T], 'single')';
fclose(fid);

fid = fopen(fullfile(ml_dir_D, 'station_ids.txt'), 'r');
station_ids = textscan(fid, '%s');
station_ids = station_ids{1};
fclose(fid);

% load observations (same as Step C)
obs = parquetread('D:\Dataset\DPastCliM-NA\verification\step_A\python\ghcn_tas_obs.parquet');
obs_jan = obs(obs.month == 1, :);
obs_jan = obs_jan(obs_jan.year >= 1875 & obs_jan.year <= 1999, :);

% build obs matrix
[unique_ids, ~, station_idx] = unique(obs_jan.ID);
[unique_years, ~, year_idx] = unique(obs_jan.year);
Y_mat = nan(length(unique_years), length(unique_ids), 'single');
for i = 1:height(obs_jan)
    Y_mat(year_idx(i), station_idx(i)) = obs_jan.value(i);
end

% load ESM time info to align years
trace_tas = 'D:\Dataset\DPastCliM-NA\TraCE21k\TraCE-21K-II.monthly.TREFHT.nc';
time_ka = double(ncread(trace_tas, 'time'));
cal_year_frac = 1950 + time_ka * 1000;
months_since_0 = round(cal_year_frac * 12);
year_arr = int32(floor(months_since_0 / 12));
month_arr = int32(mod(months_since_0, 12) + 1);
cal_mask = (year_arr >= 1875) & (year_arr <= 1999) & (month_arr == 1);
years_jan_cal = year_arr(cal_mask);

[common_years, ia, ib] = intersect(unique_years, years_jan_cal);
Y_aligned = Y_mat(ia, :);
valid_count = sum(~isnan(Y_aligned), 1);
keep = valid_count >= 30;

% match station IDs
keep_indices = find(keep); % positions of keep=true in original 11326-element unique_ids
[common_ids, ia2, ib2] = intersect(string(station_ids), unique_ids(keep));
Y_obs = Y_aligned(:, keep_indices(ib2))'; % (S, T) — map ib2 through keep_indices
Yhat = yhat(:, ia2)'; % (S, T)

% save intersection station IDs for Python comparison
fid_ids = fopen(fullfile(out_dir, 'common_station_ids.txt'), 'w');
for i = 1:length(common_ids)
    fprintf(fid_ids, '%s\n', common_ids{i});
end
fclose(fid_ids);
n_stations = size(Y_obs, 1);
fprintf('Full dataset: %d stations, %d years\n', n_stations, T);

% compute residuals
residuals = Y_obs - Yhat; % (S, T)

%% load station metadata and project
obs_meta = parquetread('D:\Dataset\DPastCliM-NA\verification\step_A\python\ghcn_tas_meta.parquet');
id_list = string(obs_meta.ID);
lat_list = obs_meta.lat;
lon_list = obs_meta.lon;

matched_station_ids = station_ids(ia2);
[~, meta_idx] = intersect(id_list, string(matched_station_ids));
st_lat = lat_list(meta_idx);
st_lon = lon_list(meta_idx);
[x_proj, y_proj] = albers_forward(st_lon, st_lat);

%% Subsample for speed
max_stations = 500;
if n_stations > max_stations
    rng(42);
    sub_idx = sort(randperm(n_stations, max_stations));
    residuals = residuals(sub_idx, :);
    x_proj = x_proj(sub_idx);
    y_proj = y_proj(sub_idx);
    n_stations = max_stations;
    fprintf('Subsampled to %d stations for SEM fitting\n', n_stations);
    % save sub_idx so Python can use the exact same subset
    fid_idx = fopen(fullfile(out_dir, 'sub_idx.bin'), 'w');
    fwrite(fid_idx, sub_idx, 'int32');
    fclose(fid_idx);
end

%% SEM fitting (following Guaita's approach)
D = sqrt((x_proj - x_proj').^2 + (y_proj - y_proj').^2);
thresholds = linspace(25000, 100000, 4);
penalty_weight = 0.1;
valid_mask = sum(~isnan(residuals), 2) >= 30;

fprintf('residuals(1,1:5): %.4f %.4f %.4f %.4f %.4f\n', residuals(1,1:5));
fprintf('valid_mask sum: %d\n', sum(valid_mask));
fprintf('NaN fraction: %.4f\n', sum(isnan(residuals(:)))/numel(residuals));
% save residuals and coords for Python comparison
fid_r = fopen(fullfile(out_dir, 'residuals_sub.bin'), 'w');
fwrite(fid_r, [n_stations, T], 'int32');
for s = 1:n_stations
    fwrite(fid_r, residuals(s,:), 'single');
end
fclose(fid_r);

fid_c = fopen(fullfile(out_dir, 'coords_sub.bin'), 'w');
fwrite(fid_c, [x_proj, y_proj], 'double');
fclose(fid_c);

best_nLL = Inf;
lambda_hat = 0.5;
W_best = eye(n_stations);
threshold_best = 50000;

for th_idx = 1:length(thresholds)
    th = thresholds(th_idx);
    h = th;
    W = exp(-(D.^2) / (2 * h^2));
    W(eye(n_stations) == 1) = 0;
    row_sums = sum(W, 2);
    row_sums(row_sums == 0) = 1;
    W = W ./ row_sums;

    eigvals = eig(W);
    rho = max(abs(eigvals));
    fprintf('  th=%d: rho=%.6f\n', th, rho);
    if rho >= 1
        W = W / (rho + 1e-2);
        fprintf('  W rescaled by %.6f\n', rho+1e-2);
    end

    % bounded optimization for lambda
    fun = @(lam) profileNLL(lam, residuals, W, valid_mask, T, penalty_weight);
    options = optimset('TolX', 1e-6, 'MaxIter', 200);
    [lam_opt, fval] = fminbnd(fun, 0, 0.999, options);

    fprintf('  threshold=%.0f, lambda=%.4f, nLL=%.2f\n', th, lam_opt, fval);

    if fval < best_nLL
        best_nLL = fval;
        lambda_hat = lam_opt;
        W_best = W;
        threshold_best = th;
    end
end

fprintf('Best: threshold=%.0f, lambda=%.6f\n', threshold_best, lambda_hat);

% debug: evaluate nLL at fixed lambda values for threshold=100000
h100 = 100000;
W100 = exp(-(D.^2)/(2*h100^2));
W100(eye(n_stations)==1) = 0;
rs = sum(W100,2); rs(rs==0)=1; W100 = W100./rs;
ev = eig(W100); rho100 = max(abs(ev));
if rho100 >= 1, W100 = W100/(rho100+1e-2); end
for test_lam = [0.1, 0.3, 0.5, 0.7, 0.738, 0.800, 0.85, 0.9]
    nll = profileNLL(test_lam, residuals, W100, valid_mask, T, penalty_weight);
    fprintf('  debug lambda=%.3f -> nLL=%.2f\n', test_lam, nll);
end

% save W_best for comparison
fid_w = fopen(fullfile(out_dir, 'W_best.bin'), 'w');
fwrite(fid_w, W_best(:), 'double');
fclose(fid_w);

%% compute eps_mat and sigma2
A_best = eye(n_stations) - lambda_hat * W_best;
eps_mat = nan(n_stations, T);
for t = 1:T
    res_t = residuals(:, t);
    valid = ~isnan(res_t);
    if sum(valid) < 2, continue; end
    A_sub = A_best(valid, valid);
    eps_mat(valid, t) = A_sub * res_t(valid);
end
sigma2_hat = sum(eps_mat.^2, 2, 'omitnan') ./ sum(~isnan(eps_mat), 2);

%% save outputs
fid = fopen(fullfile(out_dir, 'sem_params.bin'), 'w');
fwrite(fid, lambda_hat, 'double');
fwrite(fid, threshold_best, 'double');
fwrite(fid, n_stations, 'int32');
fwrite(fid, T, 'int32');
fwrite(fid, sigma2_hat, 'double');
fclose(fid);

% eps_mat: (S, T) -> write row by row
fid = fopen(fullfile(out_dir, 'eps_mat.bin'), 'w');
fwrite(fid, [n_stations, T], 'int32');
for s = 1:n_stations
    fwrite(fid, eps_mat(s, :), 'single');
end
fclose(fid);

fprintf('Step F outputs saved to: %s\n', out_dir);

%% helper functions
function nLL = profileNLL(lam, residual_mat, W, valid_mask, nTime, alpha)
    S = size(W, 1);
    A = eye(S) - lam * W;
    eps_mat = nan(S, nTime);
    for t = 1:nTime
        res_t = residual_mat(:, t);
        valid = ~isnan(res_t);
        if sum(valid) < 2, continue; end
        eps_mat(valid, t) = A(valid, valid) * res_t(valid);
    end
    sigma2 = sum(eps_mat.^2, 2, 'omitnan') ./ sum(~isnan(eps_mat), 2);
    if any(sigma2(valid_mask) <= 0)
        nLL = 1e30;
        return;
    end
    vm_idx = find(valid_mask);
    S_v = length(vm_idx);
    A_sub = A(vm_idx, vm_idx);
    sigma2_v = sigma2(vm_idx);
    if any(sigma2_v <= 0 | isnan(sigma2_v))
        nLL = 1e30;
        return;
    end
    try
        U = chol(A_sub);
        logdetA = 2 * sum(log(diag(U)));
    catch
        logdetA = log(max(abs(det(A_sub)), eps));
    end
    term1 = -(nTime/2) * sum(log(sigma2_v));
    term2 = nTime * logdetA;
    quad = 0;
    for t = 1:nTime
        et = eps_mat(vm_idx, t);
        vt = ~isnan(et);
        quad = quad + sum(et(vt).^2 ./ sigma2_v(vt));
    end
    term3 = -0.5 * quad;
    logLik = term1 + term2 + term3;
    penalty = -alpha * log(max(1 - lam, eps));
    nLL = -(logLik - penalty);
end

function [x, y] = albers_forward(lon_deg, lat_deg)
    a = 6378137.0; f = 1/298.257222101; e2 = 2*f - f^2; e = sqrt(e2);
    phi1 = 29.5*pi/180; phi2 = 45.5*pi/180; phi0 = 23*pi/180; lam0 = -96*pi/180;
    m1 = cos(phi1)/sqrt(1-e2*sin(phi1)^2);
    m2 = cos(phi2)/sqrt(1-e2*sin(phi2)^2);
    q0 = (1-e2)*(sin(phi0)/(1-e2*sin(phi0)^2) - log((1-e*sin(phi0))/(1+e*sin(phi0)))/(2*e));
    q1 = (1-e2)*(sin(phi1)/(1-e2*sin(phi1)^2) - log((1-e*sin(phi1))/(1+e*sin(phi1)))/(2*e));
    q2 = (1-e2)*(sin(phi2)/(1-e2*sin(phi2)^2) - log((1-e*sin(phi2))/(1+e*sin(phi2)))/(2*e));
    n = (m1^2-m2^2)/(q2-q1); C = m1^2 + n*q1; rho0 = a*sqrt(C-n*q0)/n;
    phi = lat_deg(:)*pi/180; lam = lon_deg(:)*pi/180;
    q = (1-e2)*(sin(phi)./(1-e2*sin(phi).^2) - log((1-e*sin(phi))./(1+e*sin(phi)))/(2*e));
    rho = a*sqrt(C-n*q)/n; theta = n*(lam-lam0);
    x = rho.*sin(theta); y = rho0 - rho.*cos(theta);
end
